Add multi-branching for experiments library

georgf opened the pull request:
Starting a draft PR here because i'll need some feedback.

So far this contains:
- all the changes to the field names in the JSON experiments list
- supporting the "branches" field in the JSON file
- cleanups / refactors that support this change.

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

georgf commented on components/service/experiments/src/test/java/mozilla/components/service/experiments/SignatureVerifierTest.kt:
```diff
@@ -80,7 +97,16 @@ class SignatureVerifierTest {
     @Test
     fun validSignatureCorrect() {
```
This is the single failing test with the current changes.

@pocmo do you remember any context around this test specifically or the signature verification code?

The failure is logged [here](https://tools.taskcluster.net/groups/WJ8WDF4tSMiMcI46LLUBOQ/tasks/ZVybi2lVQSOdkVNSYuDqYQ/runs/0/logs/public%2Flogs%2Flive.log#L6906):
> java.lang.AssertionError: expected:<true> but was:<false>
>   ...
>   at mozilla.components.service.experiments.SignatureVerifierTest.testSignature(SignatureVerifierTest.kt:1022)

Which is the assert [here](https://github.com/mozilla-mobile/android-components/pull/2702/files#diff-f708252b135a5b742cdeb293291a745dR1022):
```kotlin
assertEquals(expected, verifier.validSignature(experiments, experimentsJSON.getLong("last_modified")))
```

The changes in this PR that relates to this are mostly changing the schema of the JSON file (i.e. i changed field names and added/removed some fields), which affects [the test data](https://github.com/mozilla-mobile/android-components/pull/2702/files#diff-f708252b135a5b742cdeb293291a745dR668).
I assume... that changes the signature of the data and i should regenerate it?
Do you know how exactly?

pocmo replied:
Background: Kinto is signing the collection whenever it is changed. On the client side we can verify the signature to make sure the data we received is the correctly signed data and was not altered/intercepted/etc.

For this test code I think we took the actual data/signature from a live Kinto instance used in production, e.g.
* https://firefox.settings.services.mozilla.com/v1/buckets/fennec/collections/experiments/records
* https://firefox.settings.services.mozilla.com/v1/buckets/main/collections/focus-experiments/records

pocmo replied:
@fercarcedo Hey Fernando! Are you still around? I remember you had some great resources about the signing process - do you still have some of those links to share? :)

georgf replied:
Thanks Sebastian! I found out a bit more: Kinto signs it's experiments using the [autograph service](https://github.com/mozilla-services/autograph), i'm following up on how to get updated signatures out of that.

fercarcedo replied:
@pocmo Yes, in order to implement it I talked with Franziskus Kiefer and Julien Vehent who pointed me to [this talk](https://www.youtube.com/watch?v=b2kPo8YdLTw&t=0s&list=WL&index=2) and also to  sample implementations of the signature verification algorithm in [C++](https://searchfox.org/mozilla-central/source/security/manager/ssl/ContentSignatureVerifier.cpp) and [Go](https://github.com/mozilla-services/autograph/blob/master/tools/autograph-monitor/contentsignature.go#L35).

fercarcedo replied:
The test code is using actual data / signature taken from a live Kinto instance. The signature verification code takes the signature and the X5U certificate chain from the Kinto collection metadata URL and validates it (both of which are mocked on the test)

georgf replied:
Thanks Fernando, that makes sense, i'll go for that.

georgf replied:
The follow-up will happen in [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1543971).

georgf commented:
The changes now contain:
- The new JSON file attributes (per the [spec](https://docs.google.com/document/d/1C1k68VHKEhl1fp2oA_kkn23NZxsaloIIAJiukVQKgf0/)).
- Multi-branching support. If a client matches an experiment, they get assigned a random branch.
- Only one experiment is active at a time. That simplifies a bunch of assumptions. Introduced `ActiveExperiment` to keep that state.
- Persistence of active experiment & branch. They are stored/restored via shared prefs, so the assignment and branch stay stable.
- `withExperiment()` was updated to pass the branch name to outside consumers.
- Temporarily disabled one signing test. I'm taking this in [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1543971), but i need to sort the kinto setup first.
- Refactored library code and tests quite a bit. Among other things aggressive use of named properties for less error prone passing around of data. Also core `data` classes like `Experiment` now have all arguments required to statically make sure i'm not missing any. Test utilities loosen this up a bit, e.g. `testDefaultExperiment`.

georgf commented on components/service/experiments/src/main/java/mozilla/components/service/experiments/Experiment.kt:
```diff
@@ -4,6 +4,55 @@
 
+import android.support.annotation.VisibleForTesting
+
+@VisibleForTesting(otherwise = VisibleForTesting.NONE)
```
Yeah, that's a good point.

georgf commented on components/service/experiments/src/test/java/mozilla/components/service/experiments/KintoExperimentSourceTest.kt:
```diff
@@ -99,37 +173,64 @@ class KintoExperimentSourceTest {
+                    {
+                        "deleted": true,
+                        "id": "id",
+                        "last_modified": 1523549899999
```
These are all carried over from the previous code state.
But yes, i can see if i can fit a cleanup in. Otherwise i might file a follow-up.

georgf replied:
This is not new, it's carried over from the previous test.
I can see if i can fit a cleanup in, otherwise i might file a follow-up bug.

georgf commented on components/service/experiments/src/main/java/mozilla/components/service/experiments/ActiveExperiment.kt:
```diff
@@ -0,0 +1,61 @@
+            return ActiveExperiment(experiment, branch)
+        }
+
+        fun clear(context: Context) {
```
This should be marked `internal`, but not test-only. After #2749, we need to wire this up to experiments being stopped.
I think it's easiest if we schedule a follow-up bug, then after landing this and #2749, we go through the experiment enabling/disabling flow to make sure it works out correctly.

georgf replied:
I filed [a follow-up bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1544751) for checking the workflow.

georgf commented on components/service/experiments/src/test/java/mozilla/components/service/experiments/ExperimentTest.kt:
```diff
@@ -14,33 +14,33 @@ import org.robolectric.RobolectricTestRunner
+            schemaModified = null)
+
         assertTrue(experiment == experiment)
         assertFalse(experiment.equals(null))
```
That gives IDE warning/highlighting on this never being true.

georgf commented on components/service/experiments/src/test/java/mozilla/components/service/experiments/KintoExperimentSourceTest.kt:
```diff
@@ -99,37 +173,64 @@ class KintoExperimentSourceTest {
+                regions = listOf("US")
+            ),
+            buckets = Experiment.Buckets(5, 5),
+            lastModified = 1523549890000
```
Let's clean those up in a follow-up. Those are not new but rather carried over from the previous test code.

georgf commented:
I addressed the review feedback.

Also Travis noticed that we could use a matcher for selecting specific testing devices, so we came up with an experimental `match.debug_tags` matcher that i integrated here into the JSON schema.

georgf commented on components/service/experiments/src/main/java/mozilla/components/service/experiments/ActiveExperiment.kt:
```diff
@@ -0,0 +1,61 @@
+            context.getSharedPreferences(PREF_NAME_ACTIVE_EXPERIMENT, Context.MODE_PRIVATE)
+
+        @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+        const val PREF_NAME_ACTIVE_EXPERIMENT = "mozilla.components.service.experiments.active_experiment"
```
They are on an `internal data class` - does that actually make a difference?

georgf commented on components/service/experiments/src/main/java/mozilla/components/service/experiments/Experiment.kt:
```diff
@@ -15,85 +15,101 @@ internal data class Experiment(
+        val ratio: Int
+    )
+
+    data class Matcher(
```
That's just phrasing, it's about the what the experiment matcher/targetting requires.

georgf replied:
I'll just update the comment a bit.
